src/lib/mixer_module/actuator_test.cpp

```diff
@@ -0,0 +1,126 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2021 PX4 Development Team. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#include "actuator_test.hpp"
+
+using namespace time_literals;
+
+ActuatorTest::ActuatorTest(const OutputFunction function_assignments[MAX_ACTUATORS])
+	: _function_assignments(function_assignments)
+{
+	reset();
+}
+
+void ActuatorTest::update(int num_outputs, bool reversible_motors, float thrust_curve)
+{
+	const hrt_abstime now = hrt_absolute_time();
+
+	actuator_test_s actuator_test;
+
+	while (_actuator_test_sub.update(&actuator_test)) {
+		if (actuator_test.timestamp == 0 ||
+		    hrt_elapsed_time(&actuator_test.timestamp) > 100_ms) {
+			continue;
+		}
+
+		for (int i = 0; i < num_outputs; ++i) {
+			if ((int)_function_assignments[i] == actuator_test.function) {
+				bool in_test_mode = actuator_test.action == actuator_test_s::ACTION_DO_CONTROL;
+
+				// entering test mode?
+				if (!_in_test_mode && in_test_mode) {
+					reset();
+					_in_test_mode = true;
+				}
+
+				_output_overridden[i] = in_test_mode;
+
+				if (in_test_mode) {
+					if (actuator_test.timeout_ms > 0) {
+						_next_timeout = now + actuator_test.timeout_ms * 1000;
```

Comment:
How do we know if the conversion into `hrt_abstime` datastruct conversion happens safely without having integer overflow? (e.g. Integer fitting 1000 * timeout_ms can overflow)